File: python/src/hohmann_het/optimization.py
# Author: A Taylor | Purpose: Mission optimization solvers for HET transfers | Ref: Vallado/Curtis
"""Mission optimization module.

Provides solvers for minimizing propellant mass or burn time for
low-thrust Hall Effect Thruster orbital transfer missions.

The core algorithm is a pure-Python golden-section search, ensuring
identical numerical behavior across the tri-language implementation.
"""
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Callable

import astropy.units as u

from .propulsion import G0

# ---------------------------------------------------------------------------
# Result dataclass
# ---------------------------------------------------------------------------
@dataclass(frozen=True)
class OptimizationResult:
    """Result of a specific-impulse optimization.

    Attributes
    ----------
    optimal_isp : u.Quantity
        Optimal specific impulse [s].
    propellant_mass : u.Quantity
        Propellant mass at the optimum [kg].
    burn_time : u.Quantity
        Burn time at the optimum [s].
    objective_value : float
        Dimensionless composite objective at the optimum.
    """

    optimal_isp: u.Quantity
    propellant_mass: u.Quantity
    burn_time: u.Quantity
    objective_value: float


# ---------------------------------------------------------------------------
# Core algorithm
# ---------------------------------------------------------------------------
def golden_section_minimize(
    func: Callable[[float], float],
    a: float,
    b: float,
    tol: float = 1e-9,
) -> float:
    r"""Golden-section search for the minimum of a unimodal scalar function.

    Converges to within ``tol`` of the true minimum location.

    The interval is reduced at each step by the golden ratio:

    .. math::
        \varphi = \frac{1 + \sqrt{5}}{2} \approx 1.618

    Parameters
    ----------
    func : callable
        Scalar function :math:`f : \mathbb{R} \to \mathbb{R}` to minimize.
    a : float
        Left bound of the search interval.
    b : float
        Right bound of the search interval.
    tol : float
        Convergence tolerance on the interval width.

    Returns
    -------
    float
        Approximate location of the minimum.
    """
    gr = (math.sqrt(5.0) + 1.0) / 2.0  # golden ratio
    c = b - (b - a) / gr
    d = a + (b - a) / gr

    while abs(b - a) > tol:
        if func(c) < func(d):
            b = d
        else:
            a = c
        c = b - (b - a) / gr
        d = a + (b - a) / gr

    return (a + b) / 2.0


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------
def optimize_isp(
    m_initial: "u.Quantity | float",
    delta_v: "u.Quantity | float",
    discharge_power: "u.Quantity | float",
    anode_efficiency: float,
    isp_min: "u.Quantity | float" = 500.0,
    isp_max: "u.Quantity | float" = 5000.0,
    lambda_weight: float = 1e-4,
) -> OptimizationResult:
    r"""Optimize specific impulse to minimize a composite propellant-time objective.

    **Objective function** (Pareto trade-off between propellant mass and burn time):

    .. math::
        J(I_{sp}) = \frac{m_p(I_{sp})}{m_0}
                  + \lambda \frac{T_b(I_{sp})}{T_{\rm ref}}

    **Propellant mass** (Tsiolkovsky):

    .. math::
        m_p = m_0 \left(1 - e^{-\Delta v / (g_0 I_{sp})}\right)

    **Power-limited burn time:**

    .. math::
        T_b = \frac{m_0 \,\Delta v \, v_e}{2 \eta_a P_d}, \quad v_e = g_0 I_{sp}

    Minimized via :func:`golden_section_minimize` over
    :math:`I_{sp} \in [I_{sp,\min},\ I_{sp,\max}]`.

    Parameters
    ----------
    m_initial : astropy.units.Quantity or float
        Spacecraft initial (wet) mass [kg].
    delta_v : astropy.units.Quantity or float
        Required delta-V. If plain ``float``, assumed km/s.
    discharge_power : astropy.units.Quantity or float
        Available thruster discharge power [W].
    anode_efficiency : float
        HET anode efficiency :math:`\eta_a \in (0, 1)`.
    isp_min : astropy.units.Quantity or float
        Lower bound of Isp search [s]. Default 500 s.
    isp_max : astropy.units.Quantity or float
        Upper bound of Isp search [s]. Default 5000 s.
    lambda_weight : float
        Dimensionless weighting factor for the burn-time term.

    Returns
    -------
    OptimizationResult
        Optimal Isp and corresponding mission parameters.
    """
    if not isinstance(m_initial, u.Quantity):
        m_initial = float(m_initial) * u.kg
    if not isinstance(delta_v, u.Quantity):
        delta_v = float(delta_v) * (u.km / u.s)
    if not isinstance(discharge_power, u.Quantity):
        discharge_power = float(discharge_power) * u.W
    if not isinstance(isp_min, u.Quantity):
        isp_min = float(isp_min) * u.s
    if not isinstance(isp_max, u.Quantity):
        isp_max = float(isp_max) * u.s

    m0     = m_initial.to(u.kg).value
    dv_ms  = delta_v.to(u.m / u.s).value
    P      = discharge_power.to(u.W).value
    isp_lo = isp_min.to(u.s).value
    isp_hi = isp_max.to(u.s).value
    eta_a  = anode_efficiency
    g0     = G0.value

    if m0 <= 0.0 or dv_ms < 0.0 or P <= 0.0:
        raise ValueError("Mass, delta-V, and power must be non-negative/positive.")
    if not (0.0 < eta_a < 1.0):
        raise ValueError("Anode efficiency must be in (0, 1).")
    if isp_lo <= 0.0 or isp_hi <= isp_lo:
        raise ValueError("Isp bounds must satisfy 0 < isp_min < isp_max.")

    # Normalisation reference: burn time at minimum Isp
    ve_ref = isp_lo * g0
    T_ref  = m0 * dv_ms * ve_ref / (2.0 * eta_a * P)

    def objective(isp_s: float) -> float:
        ve      = isp_s * g0
        mp_frac = 1.0 - math.exp(-dv_ms / ve)
        t_burn  = m0 * dv_ms * ve / (2.0 * eta_a * P)
        return mp_frac + lambda_weight * (t_burn / T_ref)

    isp_opt = golden_section_minimize(objective, isp_lo, isp_hi)
    ve_opt  = isp_opt * g0
    mp_opt  = m0 * (1.0 - math.exp(-dv_ms / ve_opt))
    tb_opt  = m0 * dv_ms * ve_opt / (2.0 * eta_a * P)
    J_opt   = objective(isp_opt)

    return OptimizationResult(
        optimal_isp=isp_opt * u.s,
        propellant_mass=mp_opt * u.kg,
        burn_time=tb_opt * u.s,
        objective_value=J_opt,
    )


def min_propellant_transfer(
    h1: "u.Quantity | float",
    h2: "u.Quantity | float",
    m_initial: "u.Quantity | float",
    discharge_power: "u.Quantity | float",
    anode_efficiency: float,
    isp_min: "u.Quantity | float" = 500.0,
    isp_max: "u.Quantity | float" = 5000.0,
) -> OptimizationResult:
    """Compute Hohmann delta-V then optimize Isp for minimum propellant.

    Convenience wrapper combining :func:`~hohmann_het.dynamics.compute_hohmann`
    and :func:`optimize_isp`.

    Parameters
    ----------
    h1 : astropy.units.Quantity or float
        Initial orbit altitude [km].
    h2 : astropy.units.Quantity or float
        Final orbit altitude [km].
    m_initial : astropy.units.Quantity or float
        Spacecraft wet mass [kg].
    discharge_power : astropy.units.Quantity or float
        Available HET discharge power [W].
    anode_efficiency : float
        Anode efficiency in (0, 1).
    isp_min : astropy.units.Quantity or float
        Lower Isp search bound [s].
    isp_max : astropy.units.Quantity or float
        Upper Isp search bound [s].

    Returns
    -------
    OptimizationResult
    """
    from .dynamics import compute_hohmann

    transfer = compute_hohmann(h1, h2)
    return optimize_isp(
        m_initial=m_initial,
        delta_v=transfer.total_dv,
        discharge_power=discharge_power,
        anode_efficiency=anode_efficiency,
        isp_min=isp_min,
        isp_max=isp_max,
    )

File: matlab/+hohmann_het/Optimization.m
% Author: A Taylor | Purpose: Mission optimization solvers for HET transfers | Ref: Vallado/Curtis
classdef Optimization
% OPTIMIZATION - Mission optimization solvers for HET orbital transfers.
%
% Implements golden-section search for minimizing a composite
% propellant-mass / burn-time objective over specific impulse.
%
% The golden-section algorithm is numerically identical to the Python
% and C++ implementations, guaranteeing cross-language result parity
% within floating-point tolerance (1e-6).
%
% Usage:
%   r = hohmann_het.Optimization.optimize_isp(1000, 3852.6, 5000, 0.55)
%   r = hohmann_het.Optimization.min_propellant_transfer(400, 35786, 1000, 5000, 0.55)
%
% Author: A Taylor | Ref: Vallado/Curtis

    properties (Constant, Access = private)
        G0 = 9.80665;  % Standard gravity [m/s^2]
    end

    methods (Static)

        function x_min = golden_section_minimize(func, a, b, tol)
        % GOLDEN_SECTION_MINIMIZE  Minimize a unimodal scalar function on [a, b].
        %
        % Converges to within tol of the true minimum location.
        %
        %   phi = (1 + sqrt(5)) / 2   (golden ratio)
        %
        % Inputs:
        %   func - function handle @(x) scalar
        %   a    - left  bound of search interval
        %   b    - right bound of search interval
        %   tol  - convergence tolerance (default 1e-9)
        %
        % Output:
        %   x_min - approximate location of the minimum

            arguments
                func (1,1) {mustBeA(func, "function_handle")}
                a    (1,1) double {mustBeFinite}
                b    (1,1) double {mustBeFinite}
                tol  (1,1) double {mustBePositive} = 1e-9
            end

            gr = (sqrt(5.0) + 1.0) / 2.0;  % golden ratio
            c  = b - (b - a) / gr;
            d  = a + (b - a) / gr;

            while abs(b - a) > tol
                if func(c) < func(d)
                    b = d;
                else
                    a = c;
                end
                c = b - (b - a) / gr;
                d = a + (b - a) / gr;
            end

            x_min = (a + b) / 2.0;
        end

        function result = optimize_isp(m_initial, delta_v_m_s, discharge_power, ...
                                        anode_efficiency, isp_min, isp_max, lambda_weight)
        % OPTIMIZE_ISP  Find specific impulse minimizing composite propellant-time objective.
        %
        % Objective function:
        %
        %   J(I_sp) = m_p(I_sp)/m_0  +  lambda * T_b(I_sp)/T_ref
        %
        % Propellant mass (Tsiolkovsky):
        %
        %   m_p = m_0 * (1 - exp(-Delta_v / (g_0 * I_sp)))
        %
        % Power-limited burn time:
        %
        %   T_b = m_0 * Delta_v * v_e / (2 * eta_a * P_d),  v_e = g_0 * I_sp
        %
        % Minimized via golden-section search over I_sp in [isp_min, isp_max].
        %
        % Inputs:
        %   m_initial        - spacecraft wet mass [kg],   positive scalar
        %   delta_v_m_s      - required delta-V [m/s],     non-negative scalar
        %   discharge_power  - thruster power [W],          positive scalar
        %   anode_efficiency - eta_a in (0,1),              scalar
        %   isp_min          - lower Isp bound [s]          (default 500)
        %   isp_max          - upper Isp bound [s]          (default 5000)
        %   lambda_weight    - burn-time penalty weight     (default 1e-4)
        %
        % Output:
        %   result - struct with fields:
        %       optimal_isp      [s]
        %       propellant_mass  [kg]
        %       burn_time        [s]
        %       objective_value  [-]

            arguments
                m_initial        (1,1) double {mustBePositive,    mustBeFinite}
                delta_v_m_s      (1,1) double {mustBeNonnegative, mustBeFinite}
                discharge_power  (1,1) double {mustBePositive,    mustBeFinite}
                anode_efficiency (1,1) double {mustBeInRange(anode_efficiency, 0, 1, "exclusive", "exclusive")}
                isp_min          (1,1) double {mustBePositive,    mustBeFinite} = 500.0
                isp_max          (1,1) double {mustBePositive,    mustBeFinite} = 5000.0
                lambda_weight    (1,1) double {mustBePositive,    mustBeFinite} = 1e-4
            end

            if isp_max <= isp_min
                error('hohmann_het:Optimization:badBounds', ...
                      'isp_max must be greater than isp_min.');
            end

            G0    = hohmann_het.Optimization.G0;
            m0    = m_initial;
            dv    = delta_v_m_s;
            P     = discharge_power;
            eta_a = anode_efficiency;

            % Reference burn time at minimum Isp (for normalisation)
            ve_ref = isp_min * G0;
            T_ref  = m0 * dv * ve_ref / (2.0 * eta_a * P);

            obj = @(isp_s) ...
                (1.0 - exp(-dv / (isp_s * G0))) + ...
                lambda_weight * (m0 * dv * (isp_s * G0) / (2.0 * eta_a * P)) / T_ref;

            isp_opt = hohmann_het.Optimization.golden_section_minimize( ...
                obj, isp_min, isp_max);

            ve_opt = isp_opt * G0;
            mp_opt = m0 * (1.0 - exp(-dv / ve_opt));
            tb_opt = m0 * dv * ve_opt / (2.0 * eta_a * P);
            J_opt  = obj(isp_opt);

            result.optimal_isp     = isp_opt;
            result.propellant_mass = mp_opt;
            result.burn_time       = tb_opt;
            result.objective_value = J_opt;
        end

        function result = min_propellant_transfer(h1, h2, m_initial, ...
                                                   discharge_power, anode_efficiency, ...
                                                   isp_min, isp_max)
        % MIN_PROPELLANT_TRANSFER  Compute Hohmann dv then optimize Isp.
        %
        % Convenience wrapper combining Dynamics.compute_hohmann and optimize_isp.
        %
        % Inputs:
        %   h1, h2           - initial/final orbit altitudes [km]
        %   m_initial        - spacecraft wet mass [kg]
        %   discharge_power  - thruster power [W]
        %   anode_efficiency - eta_a in (0,1)
        %   isp_min          - lower Isp bound [s]  (default 500)
        %   isp_max          - upper Isp bound [s]  (default 5000)
        %
        % Output:
        %   result - same struct as optimize_isp

            arguments
                h1               (1,1) double {mustBePositive, mustBeFinite}
                h2               (1,1) double {mustBePositive, mustBeFinite}
                m_initial        (1,1) double {mustBePositive, mustBeFinite}
                discharge_power  (1,1) double {mustBePositive, mustBeFinite}
                anode_efficiency (1,1) double {mustBeInRange(anode_efficiency, 0, 1, "exclusive", "exclusive")}
                isp_min          (1,1) double {mustBePositive, mustBeFinite} = 500.0
                isp_max          (1,1) double {mustBePositive, mustBeFinite} = 5000.0
            end

            transfer = hohmann_het.Dynamics.compute_hohmann(h1, h2);
            dv_m_s   = transfer.total_dv * 1000.0;   % km/s -> m/s

            result = hohmann_het.Optimization.optimize_isp( ...
                m_initial, dv_m_s, discharge_power, anode_efficiency, ...
                isp_min, isp_max);
        end

    end % methods (Static)
end % classdef
